appcelerator/titanium_mobile
# [TIMOB-17363] Android/iOS: Added "MaskedImage" support to Android. Moved BLEND_MODE constants to "Ti.UI".

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-17363

**Summary:**
- Moved `BLEND_MODE_*` constants from `Ti.UI.iOS` to `Ti.UI`.
  * Constants are still defined under `Ti.UI.iOS` for backward compatibility on iOS, but are documented as deprecated.
- Added `Ti.UI.MaskedImage` support to Android.
  * Intended as alternate solution to using rounded borders around images, which has hardware acceleration issues on Android.
- Supports 9-patch images, but only if loaded from APK's "res" directory.
- Supports base View class' `tintColor` and `opacity` properties. Applied to final result.
- Super optimized masking via GPU shaders on Android for:
  * `mask` and `tint` for all blend modes.
  * `mask` and `image` for modes `SOURCE_IN` and `DESTINATION_IN`.
- All other masking combinations not in above list resort to less optimized rendering on Android.
- Android does not support the following blend modes _(will log a warning)_:
  * `BLEND_MODE_COLOR`
  * `BLEND_MODE_COLOR_BURN`
  * `BLEND_MODE_COLOR_DODGE`
  * `BLEND_MODE_DIFFERENCE`
  * `BLEND_MODE_EXCLUSION`
  * `BLEND_MODE_HARD_LIGHT`
  * `BLEND_MODE_HUE`
  * `BLEND_MODE_LUMINOSITY`
  * `BLEND_MODE_PLUS_DARKER`
  * `BLEND_MODE_SATURATION`
  * `BLEND_MODE_SOFT_LIGHT`
- Known issues _(to be written up later)_:
  * Will not auto-rotate JPEGs based on EXIF orientation.
  * `toImage()` renders bad image on Android 7.0 and higher when masking an image.

**Test:**
Run the below test procedure on...
- Android 4.1
- Android 4.4
- Android 7.0 or higher

Test procedure...
1. Create a classic app and add the contents of "MaskedImageTest.zip" to it from [TIMOB-17363](https://jira.appcelerator.org/browse/TIMOB-17363).
2. On app startup, verify appc logo is masked within a circle like [Screenshot1](https://jira.appcelerator.org/secure/attachment/65099/Screenshot1.png)
3. Tap the "Size" button and select "Ti.UI.FILL".
4. Verify image stretches to fit container like [Screenshot2](https://jira.appcelerator.org/secure/attachment/65100/Screenshot2.png).
5. Rotate device to landscape.
6. Verify image stretches to fit landscape container like [Screenshot3](https://jira.appcelerator.org/secure/attachment/65101/Screenshot3.png).
7. Rotate device back to portrait.
8. Tap the "Mode" button and select "SOURCE_OUT".
9. Verify image is only displayed outside of circle like [Screenshot4](https://jira.appcelerator.org/secure/attachment/65102/Screenshot4.png).
10. Tap the "Mask" button and select "MaskCircleLock".
11. Tap the "Image" button and select "None".
12. Tap the "Tint" button and select "Red".
13. Verify lock is red, within a white circle, and outside of circle is red like [Screenshot5](https://jira.appcelerator.org/secure/attachment/65103/Screenshot5.png).
14. Tap the "Mode" button and select "SOURCE_IN".
15. Verify lock is white with a red circle like [Screenshot6](https://jira.appcelerator.org/secure/attachment/65104/Screenshot6.png).
16. Tap the "Size" button and select "300dp".
17. Verify circle image is no longer stretched like [Screenshot7](https://jira.appcelerator.org/secure/attachment/65105/Screenshot7.png).
18. Tap the "Image" button and select "MaskVerticalGradient".
19. Verify circle has a gradient (light red on top; full red on bottom) like [Screenshot8](https://jira.appcelerator.org/secure/attachment/65106/Screenshot8.png).
